File: cpp/tools/base64_compatibility_test.cpp
/*
 * BaseFWX - Cryptography Engine
 * Copyright (C) 2020-2026  FixCraft Inc.
 * Licensed under the GNU General Public License v3.0 or later.
 */

#include "basefwx/base64.hpp"

#include <cstdint>
#include <iostream>
#include <string>
#include <vector>

int main() {
    const std::vector<std::uint8_t> values{0xFB, 0xFF};
    if (basefwx::base64::Encode(values) != "+/8=") {
        std::cerr << "FAIL: producer did not use standard Base64 alphabet\n";
        return 1;
    }

    for (const std::string encoded : {"+/8=", "-_8="}) {
        bool ok = false;
        const auto decoded = basefwx::base64::Decode(encoded, &ok);
        if (!ok || decoded != values) {
            std::cerr << "FAIL: decoder rejected alphabet: " << encoded << "\n";
            return 1;
        }
        if (!basefwx::base64::IsLikelyBase64(encoded)) {
            std::cerr << "FAIL: detector rejected alphabet: " << encoded << "\n";
            return 1;
        }
    }
    return 0;
}

File: java/src/test/java/com/fixcraft/basefwx/Base64CompatibilityTest.java
/*
 * BaseFWX - Cryptography Engine
 * Copyright (C) 2020-2026  FixCraft Inc.
 * Licensed under the GNU General Public License v3.0 or later.
 */

package com.fixcraft.basefwx;

import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class Base64CompatibilityTest {
    private static final byte[] VALUES_62_AND_63 =
            new byte[] {(byte) 0xFB, (byte) 0xFF};

    @Test
    public void emitsCanonicalStandardAlphabet() {
        assertEquals("+/8=", Base64Codec.encode(VALUES_62_AND_63));
    }

    @Test
    public void acceptsStandardAndLegacyUrlSafeAlphabets() {
        assertArrayEquals(VALUES_62_AND_63, Base64Codec.decode("+/8="));
        assertArrayEquals(VALUES_62_AND_63, Base64Codec.decode("-_8="));
        assertTrue(Base64Codec.looksLikeBase64("-_8="));
    }
}
